From c6f486cb236f8562d27cc155fcefadb65b9a9cb0 Mon Sep 17 00:00:00 2001
From: nx111 <NX111.AimH@gmail.com>
Date: Thu, 13 Dec 2018 19:18:29 +0800
Subject: [PATCH 36/42] f/b: start service foreground when it allow starup
 background after boot completed.

some old app such as samsung health auto startService faild, try
start it foreground.

Change-Id: I60888062bd2613e99789fdcc11f9e9cd20d4ec41
---
 .../com/android/server/am/ActiveServices.java   | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/services/core/java/com/android/server/am/ActiveServices.java b/services/core/java/com/android/server/am/ActiveServices.java
index 86490d9365b..a58b4b7bb2a 100644
--- a/services/core/java/com/android/server/am/ActiveServices.java
+++ b/services/core/java/com/android/server/am/ActiveServices.java
@@ -445,6 +445,23 @@ public final class ActiveServices {
             forcedStandby = true;
         }
 
+        // start service foreground when it allow starup background after boot completed.
+        if (!fgRequired) {
+            final int allowed = mAm.getAppStartModeLocked(r.appInfo.uid, r.packageName,
+                    r.appInfo.targetSdkVersion, callingPid, false, false, forcedStandby);
+            if (allowed != ActivityManager.APP_START_MODE_NORMAL) {
+                Slog.w(TAG, "Background start not allowed: service "
+                        + service + " to " + r.name.flattenToShortString()
+                        + " from pid=" + callingPid + " uid=" + callingUid
+                        + " pkg=" + callingPackage + " startFg?=" + fgRequired);
+                Slog.w(TAG, "Try start it foreground ...");
+                final int mode = mAm.mAppOpsService.checkOperation(
+                       AppOpsManager.OP_BOOT_COMPLETED, r.appInfo.uid, r.packageName);
+                if (mode == AppOpsManager.MODE_ALLOWED || mode == AppOpsManager.MODE_DEFAULT)
+                       fgRequired = true;
+            }
+        }
+
         // If this is a direct-to-foreground start, make sure it is allowed as per the app op.
         boolean forceSilentAbort = false;
         if (fgRequired) {
-- 
2.17.1

